### PR TITLE
HG-307: Ensure opt-in/req-out semantics [erlang]

### PR DESCRIPTION
**Pull request by keynslug**
http://thrift.apache.org/docs/idl#default-requiredness-implicit

### CODE REVIEW

File: compiler/cpp/src/generate/t_erlang_generator.cc [hunk 1]
```diff
@@ -940,7 +941,6 @@ void t_erlang_generator::generate_record_metadata(std::ostream& erl) {
 
   vec structs;
   gather_struct_types(get_program(), structs);
-  std::string const& ns = get_program()->get_namespace("erlang");
```
Comment (arentrue) [inline compiler/cpp/src/generate/t_erlang_generator.cc:943]: Why is this line removed? Doesn't look obvious to me.

File: lib/erlang/src/thrift_protocol.erl [hunk 2]
```diff
@@ -490,102 +494,116 @@ struct_write_loop(Proto0, [{Fid, _Req, Type, _Name, _Default} | RestStructDef],
...
         {_, _, SubType, Name, _Default} ->
-            validate(required, {SubType, Value}, [Name | Path]);
+            validate(required, {SubType, Value}, Direction, [Name | Path]);
         false ->
             throw({invalid, Path, Type, Data})
     end;
-validate(Req, {{struct, _Flavour, {Mod, Name} = Type}, Data}, Path)
-  when is_tuple(Data) ->
+validate(Req, {{struct, Flavour, {Mod, Name} = Type}, Data}, Direction, Path)
+  when Flavour /= union andalso is_tuple(Data) ->
```
Comment (arentrue) [inline lib/erlang/src/thrift_protocol.erl:563]: What's up with these additional _union_ checks?

**Comment by arentrue** [comment]
One more question. What about the _default values_ semantics in our implementation: does it follow [the specification for the default requredness case](https://thrift.apache.org/docs/idl#semantics-of-default-values)?

File: lib/erlang/src/thrift_protocol.erl (hunk 2 above)
Comment (keynslug) [reply lib/erlang/src/thrift_protocol.erl]: У нас по идее всё сводится к вот этому:

> implementation is free to write the default value anyways

**Comment by keynslug** [comment]
Well we're abusing this very clause from the spec

> implementation is free to write the default value anyways

File: lib/erlang/src/thrift_protocol.erl (hunk 2 above)
Comment (keynslug) [reply lib/erlang/src/thrift_protocol.erl:563]: If not for that additional guard an invalid union value which still looks like tuple would proceed to validation as a struct value and cause some chaos there. Which is certainly not what is expected, the validation should terminate with an error instead.

File: compiler/cpp/src/generate/t_erlang_generator.cc (hunk 1 above)
Comment (keynslug) [reply compiler/cpp/src/generate/t_erlang_generator.cc:943]: Compiler warned me of unused variable and I cut it out with no mercy.

**Comment by arentrue** [comment]
Great! They even have a specific note in the spec pointing out, that not writing a default value effectively creates protocol version dependency, since the default value assigned on a server, becomes a part of the protocol.

**Review by arentrue — APPROVED** [review]
